keymanapp/keyman
# refactor(web): OSK layout hierarchy encapsulation 📐

Yet more work toward the big OSK-core refactor (#5023).

I _was_ going to try going straight for a split of `OSKManager` into the different layout types, but after hitting several roadblocks... I realized that the current state of `VisualKeyboard` is the cause.  There's a _lot_ of layout-oriented spaghetti between `OSKManager` and `VisualKeyboard` that needs to get cleaned up if we want to successfully encapsulate different whole-OSK layout behaviors.  The internal layout (of layers, rows and keys) is significantly entangled with the external layout (floating location, size, behavior).

Part of the "fun" there... is that `VisualKeyboard` hasn't exactly encapsulated how its (internal) layout patterns work.  So, for the first step in addressing OSK layouts, this PR aims to encapsulate the existing (internal) layout logic of `VisualKeyboard` instances as-is.  So, the names of a few things and the source of a few references will shift... but this is largely a "reorganization" PR.  The most significant changes arose from dropping `.layerIndex` in favor of maintaining the current layer's ID.

Also, I did kind of hack a few older `VisualKeyboard` fields (`.lgKey`, `.hkKey`, `.spaceBar`) into properties to keep the changelog down.

------

## User Testing

@keymanapp/testers 

### Suite 1:  `desktop` form factor tests

**Platform / browser combinations**:
- [ ] Windows / Chrome
- [ ] macOS / Firefox
    - Unless otherwise specified, use the "**Test unminified KeymanWeb**" test page.

<details>
  <summary>Tests (with inlined instructions)</summary>

```
- [ ] TEST.DESKTOP.RESIZING:  Using the resize handle (<img width="28" alt="image" src="https://user-images.githubusercontent.com/25213402/126436281-2ea2c834-4fb0-45b4-b57a-8e3a9dcd817f.png">) for the desktop OSK, resize it at least three different ways.
     - The test "passes" if the OSK rescales according to the resize-handle's new position each time.  (The aspect ratio is _not_ locked, so narrow+tall and wide+short are both possible.)
- [ ] TEST.DESKTOP.LAYERS.SHIFT:  Using the `khmer_angkor` keyboard, press and hold your physical keyboard's `SHIFT` key.
    - Expected result:  on the OSK, the `shift` layer should be displayed and the `SHIFT` key highlighted.
- [ ] TEST.DESKTOP.LAYERS.CAPS:  Using the "English - English" keyboard entry on the unminified test page, do the following in sequence:
    - Press CAPS, then release.  (The OSK's CAPS key should highlight.)
    - Press and release your physical keyboard's `A` key.  (Should result in an 'a'.)
    - Click the OSK's CAPS key.  (It should no longer be highlighted.)
    - Press and release your physical keyboard's `A` key.  (Should result in an 'a' and CAPS becoming highlighted again.)
- [ ] TEST.DESKTOP.DOCS:  visit the "**Tests keyboard documentation rendering**" test page.
    - The shift layer for `lao_2008_basic` should be rendered and non-interactive.
    - Compare against the unminified test page's "Lao - Lao Basic" entry when emulating iOS for the best direct comparison against a live keyboard if desired.
```

</details>

### Suite 2:  `phone` form-factor tests (both in-app and out-of-app)

**Platform / browser combinations**:
- [ ] Chrome emulation / Android ("**Prediction - robust testing**" test page)
- [ ] Android / Keyman system keyboard
- [ ] iOS / Keyman app

<details>
  <summary>Tests (with inlined instructions)</summary>

```
- [ ] TEST.MOBILE.ROTATE.P-TO-L:
    1. With Keyman not active, load it in a portrait orientation.
    2. Once loaded, rotate the device to a landscape orientation.
        - On Android devices, you may need to press something like this for the rotation to occur:
        - The test "passes" if the OSK rotates properly.
- [ ] TEST.MOBILE.ROTATE.L-TO-P:
    1. With Keyman not active, load it in a landscape orientation.
    2. Once loaded, rotate the device to a portrait orientation.
        - On Android devices, you may need to press something like this for the rotation to occur:
        - The test "passes" if the OSK rotates properly.
- [ ] TEST.MOBILE.LAYERS.CHANGE:  Using the `sil_euro_latin` keyboard, swap among the 'default', 'shift', 'numeric', 'symbol', and 'currency' layers.
    - Expected result:  the layer should always change correctly.
- [ ] TEST.MOBILE.LAYERS.NUM_LOCK:  Using the `khmer_angkor` keyboard, swap back and forth between the 'default' and the 'numeric' layers.
    - Expected result:  there should be no "stuck highlighting" effect on the default layer's numeric key.

```

</details>

### Suite 3:  `phone` form-factor keyboard-documentation test (out-of-app only)


**Platform / browser combinations**:
- [ ] Chrome emulation / Android
- [ ] Chrome emulation / iOS
    - For both, use the "**Tests keyboard documentation rendering**" test page.

<details>
  <summary>Tests (with inlined instructions)</summary>

```
- [ ] TEST.MOBILE.DOCS.LOAD:  simply load the "**Tests keyboard documentation rendering**" test page.
    - The shift layer for `lao_2008_basic` should be rendered and non-interactive.
    - Compare against the unminified test page's "Lao - Lao Basic" entry when emulating iOS for the best direct comparison against a live keyboard if desired.
- [ ] TEST.MOBILE.DOCS.ROTATE:  The "documentation keyboard" should scale to fit the page when the device is rotated.
    - Vertical scrolling may be needed to view the whole thing in landscape; this is fine.
```

</details>
